BraveShared/Analytics/UserReferralProgram.swift

```diff
@@ -189,27 +200,64 @@ public class UserReferralProgram {
+        
+        return result
+    }
+    
+    /// Returns custom headers synchronously
+    private var customHeaders: [CustomHeaderData]? {
+        guard let customHeadersAsData = Preferences.URP.customHeaderData.value else { return nil }
+        
+        do {
+            return try NSKeyedUnarchiver.unarchiveTopLevelObjectWithData(customHeadersAsData) as? [CustomHeaderData]
+        } catch {
+            log.error("Failed to unwrap custom headers with error: \(error.localizedDescription)")
+        }
+        return nil
     }
     
     /// Checks if a custom header should be added to the request and returns its value and field.
     public class func shouldAddCustomHeader(for request: URLRequest) -> (value: String, field: String)? {
-        guard let customHeadersAsData = Preferences.URP.customHeaderData.value,
-            let customHeaders = (try? NSKeyedUnarchiver.unarchiveTopLevelObjectWithData(customHeadersAsData)) as? [CustomHeaderData],
+        guard let customHeaders = UserReferralProgram()?.customHeaders,
             let hostUrl = request.url?.host else { return nil }
         
         for customHeader in customHeaders {
             // There could be an egde case when we would have two domains withing different domain groups, that would
             // cause to return only the first domain-header it approaches.
-            for domain in customHeader.domainList {
-                if hostUrl.contains(domain) {
-                    if let allFields = request.allHTTPHeaderFields, !allFields.keys.contains(customHeader.headerField) {
-                        return (customHeader.headerValue, customHeader.headerField)
-                    }
+            for domain in customHeader.domainList where hostUrl.contains(domain) {
+                // If `domain` is "cookie only", we exclude it from HTTP Headers, and just use cookie approach
+                let cookieOnly = urpCookieOnlyDomains.contains(domain)
+                if !cookieOnly, let allFields = request.allHTTPHeaderFields, !allFields.keys.contains(customHeader.headerField) {
+                    return (customHeader.headerValue, customHeader.headerField)
                 }
             }
         }
         
         return nil
     }
+    
+    public class func insertCookies(intoStore store: WKHTTPCookieStore) {
+        guard let customHeadersDeferred = UserReferralProgram()?.fetchCustomHeaders() else { return }
+        
+        customHeadersDeferred.upon { customHeaders in
+            var cookies: [HTTPCookie] = []
+            customHeaders.forEach { header in
+                let domains = header.domainList.compactMap { URL(string: $0)?.absoluteString }
+                cookies += domains.compactMap {
+                    HTTPCookie(properties: [
+                        // Must include `.` prefix to be included in subdomains
+                        .domain: ".\($0)",
+                        .path: "/",
+                        .name: header.headerField,
```

Comment:
Ideally, it would be good to prefix the cookie name with [`__Secure-`](https://tools.ietf.org/html/draft-ietf-httpbis-rfc6265bis-02#section-4.1.3`) to enforce that the `secure` flag was set.

Reply:
We do toggle the `secure` which according to docs determines "whether the cookie should be transmitted only over secure channels." Do we still want to manually add the `__Secure-` string?

Reply:
The `__Secure-` prefix is a complement to the `secure` flag. In browsers that support it (Firefox, Chromium), the prefix prevents someone from setting a cookie with that name without the secure flag. That way the consumer of the cookie (i.e. the server application) can be assured that the cookie received was originally set with that flag and was never sent over HTTP. Without the prefix, there's no way to tell since the `secure` flag isn't part of the cookie request header.

Reply:
Gotcha, is it supported in webkit (safari)? If it won't hurt I can just add it regardless.

Reply:
Please move discussion to: https://github.com/brave/brave-ios/pull/823/files